src/background/requestHandler.js

```diff
@@ -18,72 +22,90 @@ export class RequestHandler extends Component {
    *
    * @param {*} receiver
    * @param {VPNController} controller
+   *  @param {ProxyHandler} proxyHandler
    */
   constructor(receiver, controller, proxyHandler) {
     super(receiver);
     this.controller = controller;
+    /** @type {ProxyHandler}  */
     this.proxyHandler = proxyHandler;
+    this.active = false;
+
+    propertySum(
+      controller.state,
+      proxyHandler.siteContexts,
+      (currentVPNState, currentSiteContextMap) => {
+        return resolveSiteContext(currentSiteContextMap, currentVPNState);
+      }
+    ).subscribe((proxyMap) => {
+      console.log(proxyMap);
+      this.onNewProxyMap(proxyMap);
+    });
+
     self = this;
   }
 
+  /**
+   *
+   * @param {Map<string, Array<browser.proxy.proxyInfo>>} proxyMap
+   */
+  onNewProxyMap(proxyMap) {
+    this.proxyMap = proxyMap;
+    if (proxyMap.size === 0 && this.active) {
+      this.removeRequestListener();
+      return;
+    }
+    if (!this.active && proxyMap.size > 0) {
+      this.addRequestListener();
+    }
+  }
+
   /** @type {VPNState | undefined} */
   controllerState;
+  /** @type {Map<string, SiteContext>} */
   siteContexts;
 
   async init() {
     log("Initiating RequestHandler");
-
-    this.controller.state.subscribe((s) => {
-      this.controllerState = s;
-      this.addOrRemoveRequestListener();
-    });
-
-    this.proxyHandler.siteContexts.subscribe((siteContexts) => {
-      this.siteContexts = siteContexts;
-      if (this.siteContexts.size === 0) {
-        return this.removeRequestListener();
-      }
-      return this.addOrRemoveRequestListener();
-    });
   }
 
   addRequestListener() {
-    log("Adding request listener");
-
-    browser.proxy.onRequest.addListener(this.interceptRequests, {
+    console.log(
+      `Starting listening for requests, active rules ${this.proxyMap.size}`
+    );
+    browser.proxy.onRequest.addListener(this.interceptRequests.bind(this), {
       urls: ["<all_urls>"],
     });
-  }
-
-  async addOrRemoveRequestListener() {
-    const s = this.controllerState.state;
-
-    switch (s) {
-      case "Enabled":
-        this.addRequestListener();
-        break;
-      case "Disabled":
-        this.removeRequestListener();
-        break;
-      // TODO what is the right behavior on "Unavailable"?
-    }
+    this.active = true;
   }
 
   /**
    *
-   * @param { RequestDetails } requestInfo
+   * @param { proxy.RequestDetails } requestInfo
    * @returns
    */
   async interceptRequests(requestInfo) {
-    let { url, originUrl } = requestInfo;
+    let { documentUrl } = requestInfo;
+    if (documentUrl != "https://mullvad.net/en/check") {
+      debugger;
+    }
 
-    for (let urlString of [url, originUrl]) {
```

Comment:
Curious: have you seen any requests where `documentUrl` is different than `originUrl` ? I think we'll need to keep `url` here for the situations where someone is on `aSiteWithoutProxyConfig` with requests going to `aSiteThatDoesHaveAProxyConfig`, which we'd want to reroute according to the rule.

Reply:
If you do a 3rdparty-request, yes. 
Just go to whatismyip.com and you will see requests with originURL 
```
Array [ "c.amazon-adsystem.com", undefined ]
requestHandler.js:101:17
Array [ "fonts.googleapis.com", undefined ]
requestHandler.js:101:17
Array [ "fonts.gstatic.com", undefined ]
requestHandler.js:101:17
Array [ "config.aps.amazon-adsystem.com", undefined ]
requestHandler.js:101:17
Array [ "tpc.googlesyndication.com", undefined ]
requestHandler.js:101:17
Array [ "secure.cdn.fastclick.net", undefined ]
requestHandler.js:101:17
Array [ "tags.crwdcntrl.net", undefined ]
```

> on aSiteWithoutProxyConfig with requests going to aSiteThatDoesHaveAProxyConfig,

We should not do that imho. 

``` 
evil.com (rule: Berlin)
whatIsMyIP.com (rule: None and the user is Connected to Brazil. )
```
and evil.com does `fetch(whatIsMyIP.com)` evil.com has just learned another IP the user is using right now. 


Firefox "total cookie protection" is partitions storage by top the level document origin, so i think for consistency we should do that too :)